### PR TITLE
*: sed `s,coreos/ignition,flatcar-linux/ignition,` 

### PR DESCRIPTION
**Pull request by tormath1**
In this PR, we create Go module `github.com/flatcar-linux/ignition` to support better integration Go projects.
* sed `coreos/ignition`  to `flatcar-linux/ignition` 
* run `go mod init`, `go mod tidy` and `go mod vendor`

See also: https://github.com/flatcar-linux/Flatcar/issues/410

### CODE REVIEW

File: build [hunk 1]
```diff
@@ -12,29 +12,23 @@ if [ -z ${VERSION+a} ]; then
 	VERSION=$(git describe --dirty --always)
 fi
 
-GLDFLAGS+="-X github.com/coreos/ignition/internal/version.Raw=${VERSION}"
-
-if [ ! -h gopath/src/${REPO_PATH} ]; then
-	mkdir -p gopath/src/${ORG_PATH}
-	ln -s ../../../.. gopath/src/${REPO_PATH} || exit 255
-fi
+GLDFLAGS+="-X github.com/flatcar-linux/ignition/internal/v0/version.Raw=${VERSION}"
```
Comment (invidian) [inline build]: Why `internal/v0/version` and not `v0/internal/version` ? :thinking:

Comment (tormath1) [reply build]: Ahh good catch, indeed the version is not correctly set:
```
./bin/amd64/ignition --version
Ignition was not built properly
```

it's better with the right order:
```
./bin/amd64/ignition --version
Ignition v0.34.0-45-gf7f5e498-dirty
```

:coffee:

**Comment by jepio** [comment]
Is there any way to declare the module as `github.com/flatcar-linux/ignition` but use `replace` in go.mod so that the source files don't all need to replace 'coreos' with 'flatcar-linux'? The angle I'm thinking about this from is compat with upstream and the upgrade to v2.

**Comment by tormath1** [comment]
@jepio 

> Is there any way to declare the module as `github.com/flatcar-linux/ignition` but use `replace` in go.mod so that the source files don't all need to replace 'coreos' with 'flatcar-linux'? The angle I'm thinking about this from is compat with upstream and the upgrade to v2.

I think this is something to do on the importer side with something like

```bash
require (
	github.com/coreos/ignition v0.35.0
        ...
)

replace github.com/coreos/ignition v0.35.0 => github.com/flatcar-linux/ignition/v0 v1.2.3
```

[ref](https://golang.org/ref/mod#go-mod-file-replace)

**Comment by invidian** [comment]
Note that the `replace` only works in the module itself. People importing the module will have to add their own `replace` statement, which is often not desired, as it is ignored by `go get` for example.

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+go 1.16
+
+require (
+	github.com/ajeddeloh/go-json v0.0.0-20160803184958-73d058cf8437
+	github.com/aws/aws-sdk-go v1.8.39
+	github.com/coreos/go-semver v0.1.0
+	github.com/coreos/go-systemd v0.0.0-20181031085051-9002847aa142
+	github.com/go-ini/ini v1.25.4 // indirect
+	github.com/godbus/dbus v0.0.0-20181025153459-66d97aec3384 // indirect
+	github.com/jmespath/go-jmespath v0.0.0-20160202185014-0b12d6b521d8 // indirect
```
Comment (invidian) [inline go.mod:12]: Can we do `sed -i '/indirect/d' go.mod`, it should get rid of indirect dependencies from here without side-effects.

File: build [hunk 3]
```diff
@@ -12,29 +12,23 @@ if [ -z ${VERSION+a} ]; then
 	VERSION=$(git describe --dirty --always)
 fi
 
-GLDFLAGS+="-X github.com/flatcar-linux/ignition/internal/version.Raw=${VERSION}"
-
-if [ ! -h gopath/src/${REPO_PATH} ]; then
-	mkdir -p gopath/src/${ORG_PATH}
-	ln -s ../../../.. gopath/src/${REPO_PATH} || exit 255
-fi
```
Comment (invidian) [inline build:20]: Seems like we could get rid of REPO_PATH in separate commit, so we could make the module renames atomic, so each commit compiles.

**Review by invidian — COMMENTED** [review]
Quoting https://golang.org/ref/mod#major-version-suffixes:
>Major version suffixes are not allowed at major versions v0 or v1. There is no need to change the module path between v0 and v1 because v0 versions are unstable and have no compatibility guarantee. Additionally, for most modules, v1 is backwards compatible with the last v0 version; a v1 version acts as a commitment to compatibility, rather than an indication of incompatible changes compared with v0.

> As a special case, modules paths starting with gopkg.in/ must always have a major version suffix, even at v0 and v1. The suffix must start with a dot rather than a slash (for example, gopkg.in/yaml.v2).

It looks like what we try to do here is against the Go modules documentation, so I would double check if this is something we really want to do.

Given that the Ignition packages for CAPI PR would be imported as `github.com/flatcar-linux/ignition/config/v2_3`, where `v2_3` _does not_ exist in latest version of `github.com/coreos/igntion` anymore, I still struggle to understand what's the point of adding `v0` or `v1` to the module name.

It seems CAPI concerns are mainly about the maintainability, so I think just switching to fork + creating a new tag (v0.x) on our side should be sufficient to ensure the maintainability.

If possible, I'd rather avoid doing something against Go modules guidelines, which are already confusing around `v2` for example.

BTW `github.com/flatcar-linux/ignition/v0/config/v2_3/types` double `v` in the package path is suspicious at least :smile:

File: go.mod (hunk 2 above)
Comment (tormath1) [reply go.mod:12]: Mm, only a few of them are really useless (sed + go mod tidy):
```diff
diff --git a/go.mod b/go.mod
index 73da947e..9dedb7d3 100644
--- a/go.mod
+++ b/go.mod
@@ -7,20 +7,17 @@ require (
        github.com/aws/aws-sdk-go v1.8.39
        github.com/coreos/go-semver v0.1.0
        github.com/coreos/go-systemd v0.0.0-20181031085051-9002847aa142
-       github.com/go-ini/ini v1.25.4 // indirect
-       github.com/godbus/dbus v0.0.0-20181025153459-66d97aec3384 // indirect
-       github.com/jmespath/go-jmespath v0.0.0-20160202185014-0b12d6b521d8 // indirect
+       github.com/go-ini/ini v1.63.2 // indirect
+       github.com/godbus/dbus v4.1.0+incompatible // indirect
+       github.com/jmespath/go-jmespath v0.4.0 // indirect
        github.com/pborman/uuid v0.0.0-20170612153648-e790cca94e6c
        github.com/pin/tftp v2.1.0+incompatible
-       github.com/sigma/bdoor v0.0.0-20160202064022-babf2a4017b0 // indirect
+       github.com/sigma/bdoor v0.0.0-20160624225336-b9c82b7b3c0b // indirect
        github.com/sigma/vmw-guestinfo v0.0.0-20160204083807-95dd4126d6e8
-       github.com/smartystreets/goconvey v1.7.2 // indirect
        github.com/stretchr/testify v1.7.0
        github.com/vincent-petithory/dataurl v0.0.0-20160330182126-9a301d65acbb
-       github.com/vmware/vmw-guestinfo v0.0.0-20170707015358-25eff159a728 // indirect
+       github.com/vmware/vmw-guestinfo v0.0.0-20211006225857-cc1fd90d572c // indirect
        github.com/vmware/vmw-ovflib v0.0.0-20170608004843-1f217b9dc714
        go4.org v0.0.0-20160314031811-03efcb870d84
        golang.org/x/net v0.0.0-20190320064053-1272bf9dcd53
-       golang.org/x/text v0.3.1-0.20190321115727-fe223c5a2583 // indirect
-       gopkg.in/yaml.v2 v2.4.0 // indirect
```

**Comment by tormath1** [comment]
@jepio @invidian I removed the `v0` thing then. It's now just a PR to convert `flatcar-linux/ignition` as a Go module.

**Review by jepio — APPROVED** [review]
LGTM but i don't know Go.

File: go.mod (hunk 2 above)
Comment (invidian) [reply go.mod:12]: Fair enough, I'm not sure how this works exactly either. Resolving anyway.

**Review by invidian — APPROVED** [review]
Imports fine to me, so LGTM :+1:

**Pull request merged**